kernel/boot_info/src/multiboot2.rs

```diff
@@ -79,6 +79,77 @@ impl<'a> crate::Module for &'a multiboot2::ModuleTag {
     }
 }
 
+fn kernel_memory_region(
+    boot_info: &multiboot2::BootInformation,
+) -> Result<ReservedMemoryRegion, &'static str> {
+    use crate::BootInformation;
+
+    // Our linker script specifies that the kernel will have the .init section
+    // starting at 1MB and ending at 1MB + .init size and all other kernel sections
+    // will start at (KERNEL_OFFSET + 1MB) and end at (KERNEL_OFFSET + 1MB + size).
+    // So, the start of the kernel is its physical address, but the end of it is its
+    // virtual address... confusing, I know. Thus, kernel_phys_start is the same as
+    // kernel_virt_start initially, but we remap them later in paging::init.
+    let start = PhysicalAddress::new(
+        boot_info
+            .elf_sections()?
+            .into_iter()
+            .filter(|section| section.is_allocated())
+            .map(|section| section.start_address())
+            .min()
+            .ok_or("couldn't find kernel start address")? as usize,
+    )
+    .ok_or("kernel physical start address was invalid")?;
+
+    let virtual_end = VirtualAddress::new(
+        boot_info
+            .elf_sections()?
+            .into_iter()
+            .filter(|s| s.is_allocated())
+            .map(|s| s.end_address())
+            .max()
+            .ok_or("couldn't find kernel end address")? as usize,
```

Comment:
while we're at it, let's refactor this a bit to avoid iterating over the `elf_sections()` list twice unnecessarily.